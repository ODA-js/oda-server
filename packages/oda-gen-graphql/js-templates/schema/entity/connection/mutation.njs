<#@ chunks "$$$main$$$" -#>
<#@ alias 'connection-mutations'#>
<#@ context 'entity'#>

<#- for (let connection of entity.connections.filter(f=>!f.embedded)) {
  const ctx = {
    entity,
    connection,
  }
#>

#{partial(ctx,'connection-mutations-addTo')}
#{partial(ctx,'connection-mutations-removeFrom')}

<#- } -#>