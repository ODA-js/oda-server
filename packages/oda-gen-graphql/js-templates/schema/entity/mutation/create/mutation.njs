<#@ chunks "$$$main$$$" -#>
<#@ alias 'mutations-create-mutation'#>
<#@ context 'entity'#>

<#-chunkStart(`./mutations/create/create${entity.name}.ts`); -#>
import {
  logger,
  RegisterConnectors,
  mutateAndGetPayload,
  PubSubEngine,
  Mutation,
  #{slot('import-common-mutation-create-slot')}
} from '../../../../common';
import gql from 'graphql-tag';

export default new Mutation({
  schema: gql`
    extend type RootMutation {
      create#{entity.name}(input: create#{entity.name}Input!): create#{entity.name}Payload
    }
  `,
  resolver: mutateAndGetPayload( async (args: {
    <#- for (let f of entity.args.create.args) {#>
      #{f.name}?: #{f.type},
    <#-}#>
    <#- for (let r of entity.relations) {#>
      #{r.field}?: object/*#{r.ref.entity}*/<#if(!r.single){#>[]<#}#>,
    <#-}#>
    },
    context: { connectors: RegisterConnectors, pubsub: PubSubEngine },
    info,
  ) => {
    logger.trace('create#{entity.name}');
    let create = context.connectors.#{entity.name}.getPayload(args, false);

    let result = await context.connectors.#{entity.name}.create(create);

    if (context.pubsub) {
      context.pubsub.publish('#{entity.name}', {
        #{entity.name}: {
          mutation: 'CREATE',
          node: result,
          previous: null,
          updatedFields: [],
          payload: args,
        }
      });
    }

    let #{entity.ownerFieldName}Edge = {
      cursor: result.id,
      node: result,
    };

    <#- for (let r of entity.relations.filter(f=>!f.embedded)) {#>

    if (args.#{r.field}<#if(!r.single){#> && Array.isArray(args.#{r.field}) && args.#{r.field}.length > 0<#}#> ) {
    <#if(!r.single && !r.embedded){#>
      for (let i = 0, len = args.#{r.field}.length; i < len; i++) {
    <#}#>
<#if(!r.embedded){#>
      let $item = args.#{r.field}<#if(!r.single){#>[i]<#}#> as { id,<#r.fields.forEach(f=>{#> #{f.name},<#})#> };
      if ($item) {
<#- slot('import-common-mutation-create-slot',`ensure${r.ref.entity}`) -#>
        let #{r.field} = await ensure#{r.ref.entity}({
          args: $item,
          context,
          create: true,
        });
<#}#>
<#- slot('import-common-mutation-create-slot',`link${entity.name}To${r.cField}`) -#>
        await link#{entity.name}To#{r.cField}({
          context,
          #{r.field}<#if(r.embedded){#>: args.#{r.field}<#}#>,
          #{entity.ownerFieldName}: result,
          <#-r.fields.forEach(f=>{#>
          #{f.name}: $item.#{f.name},
          <#-})#>
        });
<#-if(!r.embedded){#>
      }
<#-}#>
    <#if(!r.single && !r.embedded){#>
      }
    <#}#>
    }

    <#-}#>
    return {
      #{entity.ownerFieldName}: #{entity.ownerFieldName}Edge,
    };
  }),
});
