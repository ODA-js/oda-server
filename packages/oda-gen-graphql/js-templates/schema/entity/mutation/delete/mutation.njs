<#@ chunks "$$$main$$$" -#>
<#@ alias 'mutations-delete-mutation'#>
<#@ context 'entity'#>

<#-chunkStart(`./mutations/delete/delete${entity.name}.ts`); -#>

import {
  logger,
  RegisterConnectors,
  mutateAndGetPayload,
  PubSubEngine,
  Mutation,
  #{slot('import-common-mutation-delete-slot')}
} from '../../../../common';
import gql from 'graphql-tag';

export default new Mutation({
  schema: gql`
    extend type RootMutation {
      delete#{entity.name}(input: delete#{entity.name}Input!): delete#{entity.name}Payload
    }
  `,
  resolver: mutateAndGetPayload( async (args: {
   <#- for (let f of entity.args.remove.args) {#>
      #{f.name}?: #{f.type},
    <#-}#>
    <#- for (let f of entity.complexUnique) {
      let args = `${f.fields.map(f=>`${f.name}?: ${f.type}`).join(', ')}`;
      #>
      // #{f.name}
      #{args},
    <#-}#>
    },
    context: {
      connectors: RegisterConnectors,
      pubsub: PubSubEngine,
      userGQL: (args: any)=>Promise<any> },
    info,
  ) => {
    logger.trace('delete#{entity.name}');
    let result;
    if (args.id) {
<# slot('import-common-mutation-delete-slot',`unlink${entity.name}FromAll`) -#>
      await unlink#{entity.name}FromAll([{
        key: 'id',
        type: 'ID',
        value: args.id,
      }],
        context,
      );

      result = await context.connectors.#{entity.name}.findOneByIdAndRemove(args.id);
    <#- for (let f of entity.args.remove.find) {#>
    } else if (args.#{f.name}) {
<# slot('import-common-mutation-delete-slot',`unlink${entity.name}FromAll`) -#>
      await unlink#{entity.name}FromAll([{
        key: '#{f.name}',
        type: '#{f.gqlType}',
        value: args.#{f.name},
      }],
        context,
      );

      result = await context.connectors.#{entity.name}.findOneBy#{f.cName}AndRemove(args.#{f.name});
    <#-}#>
    <#- for (let f of entity.complexUnique) {
      let findBy = f.fields.map(f=>f.uName).join('And');
      let loadArgs = `${f.fields.map(f=>`args.${f.name}`).join(', ')}`;
      let condArgs = `${f.fields.map(f=>`args.${f.name}`).join(' && ')}`;
      #>
    } else if (#{condArgs}) {
<# slot('import-common-mutation-delete-slot',`unlink${entity.name}FromAll`) -#>
      await unlink#{entity.name}FromAll([
        <#-f.fields.forEach((f, i)=>{#>
        <#-if(i>0){#>, <#}#>{
          key: '#{f.name}',
          type: '#{f.gqlType}',
          value: args.#{f.name},
        }
        <#-});-#>],
        context,
      );

      result = await context.connectors.#{entity.name}.findOneBy#{findBy}AndRemove(#{loadArgs});
    <#-}#>
    }

    if (!result) {
      throw new Error('item of type #{entity.name} is not found for delete');
    }

    if (context.pubsub) {
      context.pubsub.publish('#{entity.name}', {
        #{entity.name}: {
          mutation: 'DELETE',
          node: result,
          previous: null,
          updatedFields: [],
          payload: args,
        }
      });
    }

    return {
      deletedItemId: result.id,
      #{entity.ownerFieldName}: result,
    };
  }),
})
