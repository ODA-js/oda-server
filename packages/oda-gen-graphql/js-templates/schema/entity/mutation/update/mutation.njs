<#@ chunks "$$$main$$$" -#>
<#@ alias 'mutations-update-mutation'#>
<#@ context 'entity'#>

<#-chunkStart(`./mutations/update/update${entity.name}.ts`); -#>

import {
  logger,
  RegisterConnectors,
  mutateAndGetPayload,
  PubSubEngine,
  Mutation,
  #{slot('import-common-mutation-update-slot')}
} from '../../../../common';
import gql from 'graphql-tag';
import { merge } from 'lodash';

export default new Mutation({
  schema: gql`
    extend type RootMutation {
      update#{entity.name}(input: update#{entity.name}Input!): update#{entity.name}Payload
    }
  `,
  resolver: mutateAndGetPayload( async (args: {
    <#- for (let f of entity.args.update.args) {#>
      #{f.name}?: #{f.type},
    <#-}#>
    <#- for (let r of entity.relations) {#>
      #{r.field}?: object/*#{r.ref.entity}*/<#if(!r.single){#>[]<#}#>,
      #{r.field}Unlink?: object/*#{r.ref.entity}*/<#if(!r.single){#>[]<#}#>,
      #{r.field}Create?: object/*#{r.ref.entity}*/<#if(!r.single){#>[]<#}#>,
    <#-}#>
    },
    context: { connectors: RegisterConnectors, pubsub: PubSubEngine },
    info,
  ) => {
    logger.trace('update#{entity.name}');
    let payload = context.connectors.#{entity.name}.getPayload(args);

    let result;
    let previous;

    if (args.id) {
      previous = await context.connectors.#{entity.name}.findOneById(args.id);
      result = await context.connectors.#{entity.name}.findOneByIdAndUpdate(args.id, merge({}, previous, payload),);
    <#- for (let f of entity.args.update.find) {#>
    } else if (args.#{f.name}) {
      delete payload.#{f.name};
      previous = await context.connectors.#{entity.name}.findOneBy#{f.cName}(args.#{f.name});
      result = await context.connectors.#{entity.name}.findOneBy#{f.cName}AndUpdate(args.#{f.name}, merge({}, previous, payload),);
    <#-}#>
    <#- for (let f of entity.complexUnique) {
      let findBy = f.fields.map(f=>f.uName).join('And');
      let loadArgs = `${f.fields.map(f=>`args.${f.name}`).join(', ')}`;
      let condArgs = `${f.fields.map(f=>`args.${f.name}`).join(' && ')}`;
    #>
    } else if (#{condArgs}) {
      <#-for(let fn of f.fields){#>
      delete payload.#{fn.name};
      <#-}#>
      previous = await context.connectors.#{entity.name}.findOneBy#{findBy}(#{loadArgs});
      result = await context.connectors.#{entity.name}.findOneBy#{findBy}AndUpdate(#{loadArgs}, merge({}, previous, payload),);
    <#-}#>
    }

    if (!result) {
      throw new Error('item of type #{entity.name} is not found for update');
    }

    if (context.pubsub) {
      context.pubsub.publish('#{entity.name}', {
        #{entity.name}: {
          mutation: 'UPDATE',
          node: result,
          previous,
          updatedFields: Object.keys(payload).filter(f => payload[f]!== undefined),
          payload: args,
        }
      });
    }

    <#- for (let r of entity.relations.filter(f=>!f.embedded)) {#>
    if (args.#{r.field}Unlink<#if(!r.single){#> && Array.isArray(args.#{r.field}Unlink) && args.#{r.field}Unlink.length > 0<#}#> ) {
    <#if(!r.single){#>
      for (let i = 0, len = args.#{r.field}Unlink.length; i < len; i++) {
    <#}#>
      let $item = args.#{r.field}Unlink<#if(!r.single){#>[i]<#}#>;
      if ($item) {
<# slot('import-common-mutation-update-slot',`ensure${r.ref.entity}`) -#>
        let #{r.field} = await ensure#{r.ref.entity}({
          args: $item,
          context,
          create: false,
        });
<# slot('import-common-mutation-update-slot',`unlink${entity.name}From${r.cField}`) -#>
        await unlink#{entity.name}From#{r.cField}({
          context,
          #{r.field},
          #{entity.ownerFieldName}: result,
        });
      }
    <#if(!r.single){#>
      }
    <#}#>
    }

    if (args.#{r.field}Create<#if(!r.single){#> && Array.isArray(args.#{r.field}Create) && args.#{r.field}Create.length > 0<#}#> ) {
    <#if(!r.single){#>
      for (let i = 0, len = args.#{r.field}Create.length; i < len; i++) {
    <#}#>
      let $item = args.#{r.field}Create<#if(!r.single){#>[i]<#}#> as { id,<#r.fields.forEach(f=>{#> #{f.name},<#})#> };
      if ($item) {
<# slot('import-common-mutation-update-slot',`ensure${r.ref.entity}`) -#>
        let #{r.field} = await ensure#{r.ref.entity}({
          args: $item,
          context,
          create: true,
        });

<# slot('import-common-mutation-update-slot',`link${entity.name}To${r.cField}`) -#>
        await link#{entity.name}To#{r.cField}({
          context,
          #{r.field},
          #{entity.ownerFieldName}: result,
          <#-r.fields.forEach(f=>{#>
          #{f.name}: $item.#{f.name},
          <#-})#>
        });
      }
    <#if(!r.single){#>
      }
    <#}#>
    }

    if (args.#{r.field}<#if(!r.single){#> && Array.isArray(args.#{r.field}) && args.#{r.field}.length > 0<#}#> ) {
    <#if(!r.single){#>
      for (let i = 0, len = args.#{r.field}.length; i < len; i++) {
    <#}#>
      let $item = args.#{r.field}<#if(!r.single){#>[i]<#}#> as { id,<#r.fields.forEach(f=>{#> #{f.name},<#})#> };
      if ($item) {

<# slot('import-common-mutation-update-slot',`ensure${r.ref.entity}`) -#>
        let #{r.field} = await ensure#{r.ref.entity}({
          args: $item,
          context,
          create: false,
        });

<# slot('import-common-mutation-update-slot',`link${entity.name}To${r.cField}`) -#>
        await link#{entity.name}To#{r.cField}({
          context,
          #{r.field},
          #{entity.ownerFieldName}: result,
          <#-r.fields.forEach(f=>{#>
          #{f.name}: $item.#{f.name},
          <#-})#>
        });
      }
    <#if(!r.single){#>
      }
    <#}#>
    }

    <#-}#>
    return {
      #{entity.ownerFieldName}: result,
    };
  }),
});
