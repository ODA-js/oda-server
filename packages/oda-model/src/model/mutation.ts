import {
  MutationInput,
  MutationStorage,
  MutationMeta,
  MapToHash,
  HashToMap,
  MetaModelType,
} from './interfaces';
import { ModelBase } from './modelbase';
import { IMutation } from './interfaces/model';

export class Mutation
  extends ModelBase<MutationMeta, MutationInput, MutationStorage>
  implements IMutation {
  public modelType: MetaModelType = 'mutation';
  constructor(inp: MutationInput) {
    super(inp);
  }
  public get args() {
    return this.$obj.args;
  }

  public get payload() {
    return this.$obj.payload;
  }

  public updateWith(obj: MutationInput) {
    if (obj) {
      super.updateWith(obj);
      this.$obj.args = HashToMap(obj.args);
      this.$obj.payload = HashToMap(obj.payload);
    }
  }

  public toObject(): MutationInput {
    return {
      ...super.toObject(),
      args: MapToHash(this.$obj.args),
      payload: MapToHash(this.$obj.payload),
    };
  }
}
