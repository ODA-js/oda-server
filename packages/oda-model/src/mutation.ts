import { merge } from 'lodash';
import { ElementMetaInfo, Internal } from './element';
import {
  IFieldArgs,
  AsHash,
  MetaModelType,
  HashToMap,
  Nullable,
  assignValue,
  NamedArray,
  ArrayToMap,
  MapToArray,
} from './types';
import {
  IModelBase,
  ModelBaseInternal,
  ModelBaseInput,
  ModelBase,
  ModelBaseOutput,
} from './modelbase';
import decapitalize from './lib/decapitalize';

export interface IMutation
  extends IModelBase<MutationMetaInfo, MutationInput, MutationOutput> {
  /**
   * set of arguments
   */
  readonly args: Map<string, IFieldArgs>;
  /**
   * set of output fields
   */
  readonly payload: Map<string, IFieldArgs>;
}

export interface MutationMetaInfo extends ElementMetaInfo {
  acl: {
    /** if packages allowed to execute mutation */
    execute: string[];
  };
}

export interface MutationInternal extends ModelBaseInternal {
  args: Map<string, IFieldArgs>;
  payload: Map<string, IFieldArgs>;
}

export interface MutationInput extends ModelBaseInput<MutationMetaInfo> {
  args: AsHash<IFieldArgs> | NamedArray<IFieldArgs>;
  payload: AsHash<IFieldArgs> | NamedArray<IFieldArgs>;
}

export interface MutationOutput extends ModelBaseOutput<MutationMetaInfo> {
  args: NamedArray<IFieldArgs>;
  payload: NamedArray<IFieldArgs>;
}

export const mutationDefaultMetaInfo = { acl: { execute: [] } };
export const mutationDefaultInput = { metadata: mutationDefaultMetaInfo };

export class Mutation
  extends ModelBase<
    MutationMetaInfo,
    MutationInput,
    MutationInternal,
    MutationOutput
  >
  implements IMutation {
  public get modelType(): MetaModelType {
    return 'mutation';
  }

  constructor(init: MutationInput) {
    super(merge({}, mutationDefaultInput, init));
  }

  public get args(): Map<string, IFieldArgs> {
    return this[Internal].args;
  }

  public get payload(): Map<string, IFieldArgs> {
    return this[Internal].payload;
  }

  public updateWith(input: Nullable<MutationInput>) {
    super.updateWith(input);

    assignValue<MutationInternal, MutationInput, string>({
      src: this[Internal],
      input,
      field: 'name',
      effect: (src, value) => (src.name = decapitalize(value.trim())),
      required: true,
    });

    assignValue<MutationInternal, MutationInput, MutationInput['args']>({
      src: this[Internal],
      input,
      field: 'args',
      effect: (src, value) =>
        (src.args = Array.isArray(value)
          ? ArrayToMap(value)
          : HashToMap(value)),
      required: true,
      setDefault: src => (src.args = new Map()),
    });

    assignValue<MutationInternal, MutationInput, MutationInput['payload']>({
      src: this[Internal],
      input,
      field: 'payload',
      effect: (src, value) =>
        (src.payload = Array.isArray(value)
          ? ArrayToMap(value)
          : HashToMap(value)),
      required: true,
    });
  }

  public toObject(): MutationOutput {
    return merge({}, super.toObject(), {
      args: MapToArray(this[Internal].args, (name, value) => ({
        ...value,
        name,
      })),
      payload: MapToArray(this[Internal].payload, (name, value) => ({
        ...value,
        name,
      })),
    } as Partial<MutationOutput>);
  }
}
