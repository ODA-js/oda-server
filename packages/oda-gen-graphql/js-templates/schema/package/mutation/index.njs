<#@ chunks "$$$main$$$" -#>
<#@ context 'pkg' #>
<#@ alias 'mutations/index' #>

<#- chunkStart(`./mutations/index.ts`); -#>
<# pkg.mutations.forEach(s=>{#>
import #{s.name} from './#{s.name}';
<#})#>
import { Schema } from '../common';

export default new Schema ({
  name: '#{pkg.name}.mutations',
items:[<# pkg.mutations.forEach(s=>{#>
 #{s.name},
<#})#>],
});


<# pkg.mutations.forEach(s=>{#>
<#- chunkStart(`./mutations/${s.name}/index.ts`); -#>
import { Schema } from '../../common';
import gql from 'graphql-tag';
import mutation from './mutation';
import input from './input';
import payload from './payload';

export default new Schema({
  name: '#{s.name}.mutation',
  items:[mutation, input, payload],
})

<#- chunkStart(`./mutations/${s.name}/mutation.ts`); -#>
import { Mutation, logger, mutateAndGetPayload } from '../../common';
import gql from 'graphql-tag';

export default new Mutation({
  schema:gql`
    extend type RootMutation {
      #{s.name}(input: #{s.name}Input!): #{s.name}Payload
    }
  `,
  resolver: mutateAndGetPayload(
    async (
    args: {
  <#- for (let f of s.args) {#>
      #{f.name}?: #{f.type.ts},
  <#-}#>
    },
    context,
    info
  ) => {
    logger.trace('#{s.name}');
    let result: {
      // what must be in output
  <#- for (let f of s.payload) {#>
      #{f.name}?: any; // #{f.type.ts},
  <#-}#>
    };
    result = {};
    // put your code here
    return result;
  })
})

<#- chunkStart(`./mutations/${s.name}/input.ts`); -#>

import { Input } from '../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
    input #{s.name}Input {
    <# for(let arg of s.args){-#>
      #{arg.name}: #{arg.type.gql}
    <#}-#>
    }
  `,
});



<#- chunkStart(`./mutations/${s.name}/payload.ts`); -#>
import { Type } from '../../common';
import gql from 'graphql-tag';

export default new Type({
  schema: gql`
    type #{s.name}Payload {
    <# for(let arg of s.payload){-#>
      #{arg.name}: #{arg.type.gql}
    <#}-#>
    }
  `,
});

<#})#>
