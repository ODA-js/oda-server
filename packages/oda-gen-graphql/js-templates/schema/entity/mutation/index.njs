<#@ chunks "$$$main$$$" -#>
<#@ alias 'mutation-index'#>
<#@ context 'ctx'#>

<#-chunkStart(`./mutations/index.ts`); -#>
import create from './create';
import _delete from './delete';
import update from './update';
import embed#{ctx.types.name}Input from './embed#{ctx.types.name}Input';

import { Schema } from '../../../common';

export default new Schema({
  name: '#{ctx.types.name}.mutations',
  items: [create, _delete, update, embed#{ctx.types.name}Input],
});
<#- chunkEnd(); -#>

#{partial(ctx.types, 'mutations-embed')}
#{partial(ctx, 'mutations-create-index')}
#{partial(ctx, 'mutations-delete-index')}
#{partial(ctx, 'mutations-update-index')}



