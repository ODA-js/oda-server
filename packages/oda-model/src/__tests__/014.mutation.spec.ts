import 'jest';
import { Mutation, mutationDefaultMetaInfo } from '../mutation';

describe('Mutation', () => {
  it('default', () => {
    const res = new Mutation({
      name: 'LoginUser',
      args: [{ name: 'login' }, { name: 'password' }],
      payload: [{ name: 'token' }],
    });
    expect(res.name).toBe('loginUser');
    expect(res.metadata).toMatchObject(mutationDefaultMetaInfo);
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
});
