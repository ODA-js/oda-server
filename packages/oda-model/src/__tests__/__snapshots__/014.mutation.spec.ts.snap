// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation default: toObject 1`] = `
Object {
  "args": Array [
    Object {
      "name": "login",
    },
    Object {
      "name": "password",
    },
  ],
  "description": undefined,
  "metadata": Object {
    "acl": Object {
      "execute": Array [],
    },
  },
  "name": "loginUser",
  "payload": Array [
    Object {
      "name": "token",
    },
  ],
  "title": undefined,
}
`;
